### PR TITLE
Added initialization logic when changing container attributes

### PR DESCRIPTION
**Pull request by devxom**
Added logic to manual update the Likely instance when changing the data attributes of the container on which Likely is initialized

**Changes:**
- Added the `updateOptionFromDataSet` method to update the Likely instance configuration after from the data attributes after initialization
- Added an html page for the ability to manually and automatically test data updates when changing data attributes
- Added automatic tests to verify that after changing the data-url attribute on the widget container when opening the model window for sharing, the updated url from the data-url attribute on the container is used
- Added an example to README.md to use the logic for updating the configuration from the date of attributes after initializing the Likely instance

Fixed: #151 

<details>
<summary>Previous version of description</summary>
Added logic to initialize the library when changing the data attributes of the container on which Likely is initialized

**Changes:**
- When initializing a shared sharing widget, the creation of MutationObserver was added to change the attributes of the widget container
- Added an html page for the ability to manually and automatically test data updates when changing data attributes
- Added automatic tests to verify that after changing the data-url attribute on the widget container when opening the model window for sharing, the updated url from the data-url attribute on the container is used

**Important:** to determine that the data-url attribute on the container has changed, it uses the [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) API which works on a slightly narrower number of browsers. **In browsers without support for this API, just this logic will not work.**

img

Fixed: #151 
</details>

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -339,6 +339,38 @@ describe('Likely', function () {
...
+        });
+
+
+        const testedServices = [
+            { name: 'Facebook', likelyName: 'facebook', urlRegex: `${encodeURIComponent(targetUrl)}` },
+            { name: 'Odnoklassniki', likelyName: 'odnoklassniki', urlRegex: `${encodeURIComponent(targetUrl)}` },
+            { name: 'Telegram', likelyName: 'telegram', urlRegex: `${encodeURIComponent(targetUrl)}` },
+            { name: 'Twitter', likelyName: 'twitter', urlRegex: `${encodeURIComponent(targetUrl)}` },
+            { name: 'VK', likelyName: 'vkontakte', urlRegex: `${encodeURIComponent(targetUrl)}` },
+            // Pinterest and Linkedin do not contain sharing URLs if not authorized and the test will fail
```
Comment (vitkarpov) [inline test/index.js]: Just out of curiosity, wouldn't this happen for these buttons in production?

**Comment by vitkarpov** [comment]
> In browsers without support for this API, just this logic will not work.

Could users initialize the widget once again manually?

**Comment by NikolayRys** [comment]
Hi, @devxom, thank you for the PR, I appreciate your effort.

However, after some consideration, I've got a feeling that the proposed solution might be unnecessarily complicated. Namely, I think it's sufficient to provide the client with something like `likely.reinitialize` or `likely.update_options` to use when the attributes are changed.

If the user needs to detect it automatically, they can add a Mutation Observer approach on their own. In the scope of your pull request, I suggest to keep the content of "updateOptionFromDataSet" and remove the observer.

What do you think?

File: test/index.js (hunk 1 above)
Comment (devxom) [reply test/index.js]: This is a problem only for this particular auto test. We cannot determine that the data has been verified. It's not a problem.

**Comment by devxom** [comment]
@NikolayRys, i agree. Adding DOM attribute tracking logic is unnecessarily complicated in the library itself.

Within a few days I will update the pull request. Where from the edits will be:
- logic updateOptionFromDataSet
- tests
- an example in README.md that will show how, for example, you can use MutationsObserver from outside the library

**Comment by NikolayRys** [comment]
Sounds awesome, thank you.

**Comment by devxom** [comment]
Updated patch and description for pull request. Review please.

File: readme.md [hunk 2]
```diff
@@ -163,6 +163,41 @@ You can set `data-image` and `data-description` attributes to set up an image an
...
+    <div class="pinterest">Pin</div>
+    <div class="telegram">Send</div>
+    <div class="linkedin">Share</div>
+    <div class="whatsapp">Send</div>
+</div>
+```
+
+3. Some logic for update data attributes
+
+2. Logic for update configuration on instance *Likely*
```
Comment (vitkarpov) [inline readme.md]: It's weird but the list is rendered as 1, 3, 4 (check out the preview on GitHub) 🤔 
I don't get the second item, "some logic to update attributes", could you elaborate, pls?

File: readme.md [hunk 3]
```diff
@@ -163,6 +163,41 @@ You can set `data-image` and `data-description` attributes to set up an image an
...
+
+```javascript
+// find in dom likely
```
Comment (vitkarpov) [inline readme.md]: Probably it's better to call it "the DOM"? (like developers.mozilla [does](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction))

File: source/widget.js [hunk 4]
```diff
@@ -37,6 +37,13 @@ class Likely {
         }
     }
 
+    updateOptionFromDataSet() {
```
Comment (vitkarpov) [inline source/widget.js]: Shouldn't it be plural, I mean updateOption**s**FromDataSet?

**Comment by devxom** [comment]
> @devxom @NikolayRys guys, what do you think about to not introduce a new method but use existing `init`? It's like there's only one way to (re)initialize the widget.

It sounds logical, I tested it works well. I updated the edits leaving only an example and tests.

File: readme.md (hunk 2 above)
Comment (devxom) [reply readme.md]: I will rewrite the example so that it displays normally.

**Comment by vitkarpov** [comment]
@devxom sorry, I'm not sure I get it right. Are you going to change the code in a way it uses `init` only or leave it as-is? (the latter is fine, just asking)

**Comment by devxom** [comment]
@vitkarpov I liked the idea of such a simplification, I will update the edits in the near future using `init`

**Comment by vitkarpov** [comment]
Gotha, thanks! 👍

File: test/utils/getLikelyPage.js [hunk 5]
```diff
@@ -5,6 +5,7 @@ const LikelyPage = {
     NO_AUTOINIT: 'no-autoinit.html',
     NO_AUTOINIT_MULTIPLE: 'no-autoinit-multiple.html',
     ISSUE_67: 'issues/67.html',
+    ISSUE_151: 'issues/151.html',
```
Comment (NikolayRys) [inline test/utils/getLikelyPage.js]: Those names for the tests and pages aren't great. Let's please use some more descriptive as `UPDATE_OPTIONS: 'update-options.html'`.
I will eventually refactor that 61 too.

**Comment by NikolayRys** [comment]
@devxom @vitkarpov Guys, I need your help. 
I see that we have a test that covers changing attributes with reinitialization after that: https://github.com/NikolayRys/Likely/blob/master/test/index.js#L158

Could you help me figure out, does this pull request add any capabilities beyond that?

**Comment by vitkarpov** [comment]
@NikolayRys What do you mean? What capabilities exactly?

**Comment by NikolayRys** [comment]
I'm under impression that we already have a completely functional re-initialization logic in place, could you please confirm/deny this? 
If it is indeed the case, then do we still need the separate method from this PR that updates attributes from the HTML tag?

**Comment by vitkarpov** [comment]
Yeah, I see what you mean. It seems that this PR doesn't do anything new 🤔I think we need to reproduce what  #151 reports, probably it's already fine.

**Comment by devxom** [comment]
Really necessary logic to update configuration from date attributes. In this form, Pull Request does not make sense.

@NikolayRys @vitkarpov 

How do you look at simply describing the approach to do when using dynamic date attributes and adding a test that checks and shows how this happens?

**Comment by vitkarpov** [comment]
Sounds good to me! 👍
-- 
Cheers,
Viktor

**Comment by NikolayRys** [comment]
Ee definitely lacked this information about that in the readme, so at least this part of work in this PR can be kept.
 
@devxom But about the test - doesn't this test: https://github.com/NikolayRys/Likely/blob/master/test/index.js#L158 already cover it?

**Comment by devxom** [comment]
@NikolayRys 

> @devxom But about the test - doesn't this test: https://github.com/NikolayRys/Likely/blob/master/test/index.js#L158 already cover it?

You are right, indeed this test covers this case

This weekend I will update the Pull Request. I will remove unnecessary edits and update changes with edits in README.md with an explanation of how to reinitialize Likely when using dynamic date attributes

**Comment by NikolayRys** [comment]
@devxom No update yet?

**Comment by devxom** [comment]
@NikolayRys Reassembled edits based on comments. Please take another look at the code.

**Comment by NikolayRys** [comment]
Hi, @devxom, thanks for taking care of this.

**Pull request merged**